### ISSUE TITLE
Import Attributes

### ISSUE THREAD

**Issue by xtuc** [mozilla/standards-positions#373]
## Request for Mozilla Position on an Emerging Web Specification

* Specification Title: Import attributes
* Specification or proposal URL: https://github.com/tc39/proposal-import-attributes
* Caniuse.com URL (optional): 
* Bugzilla URL (optional): 
* Mozillians who can provide input (optional): 

### Other information

Currently Stage 2.

**Comment by dbaron** [comment]
cc @codehag @annevk @jorendorff @tschneidereit @linclark

**Comment by hober** [comment]
Note that this feature (or something like it) is necessary to move forward on HTML/CSS/JSON modules.

**Comment by tschneidereit** [comment]
Also CC @lukewagner. Does it make sense to think about how this interacts with [Wasm module linking](https://github.com/WebAssembly/module-linking)?

**Comment by xtuc** [comment]
The proposal has a [WebAssembly section](https://github.com/tc39/proposal-import-attributes#webassembly). It wouldn't be in TC39 but our plan is to allow to express the same attributes than in JavaScript. The Wasm producer will need some syntax, or can based on the file extension.

**Comment by lukewagner** [comment]
The use of a custom section to allow wasm modules imported via ESM-integration to import non-wasm/JS modules seems like a reasonable approach.  Other than that, there's not any interaction I can think of with wasm module linking until "evaluator attributes" are added (which [it sounds like is explicitly postponed](https://github.com/tc39/proposal-import-attributes#restriction-to-check-attributes)).  So :thumbsup: from a wasm pov.

**Comment by codehag** [comment]
I discussed this with @annevk some time ago - it's worth pursuing with a few caveats. However, those have been met before stage 2 (we considered inbound syntax as important, as otherwise it might not be used). I don't think our support has changed since. :+1:

**Comment by littledan** [comment]
@lukewagner Do you think "evaluator attributes" would be useful for expressing permissions/capabilities for Wasm modules?

**Comment by annevk** [comment]
@xtuc note that Wasm should not use the file externsion on the web as that would be counter to how the web works.

**Comment by lukewagner** [comment]
@littledan I think so: evaluator attributes would allow JS code to pass the import values used to instantiate a wasm module (either to limit capabilities or to allow explicit parameterization) while still getting to use the ESM module loader system.  To do so without evaluator attributes, you need to resort to using the imperative JS API and manual `fetch()`ing.

**Comment by dbaron** [comment]
So does it make sense to mark this as `worth prototyping`, or are there things that should happen before we're willing to do that?

**Comment by codehag** [comment]
I think we can mark it as `worth prototyping`

**Comment by dandclark** [comment]
This proposal was renamed to "Import Assertions" to reflect the keyword we ended up settling on (`assert`), and to reflect that the presence of the attributes won't affect the behavior of a module, only block it from loading (as opposed to potential "evaluator attributes").

The old URL in activities.json will redirect to the [renamed proposal](https://github.com/tc39/proposal-import-assertions) so it's probably not critical to change anything, but we just wanted to keep everyone updated.

**Comment by dandclark** [comment]
At the time that this position request was originally marked `worth prototyping`, JSON modules was part of the import attributes proposal. Since then, JSON modules has been split off into a separate proposal: https://github.com/tc39/proposal-json-modules
(This was decided here: https://github.com/tc39/notes/blob/master/meetings/2020-07/july-21.md#conclusion--resolution-2).  The actual behavior of JSON modules has not changed in any significant way.

Because of this split, the position at https://github.com/mozilla/standards-positions/commit/da631527be5346a8993496e9d9f1cf36219bde12 is now ambiguous; it references JSON modules but the spec it links to no longer contains JSON modules.

Would it be possible to update this position to include https://github.com/tc39/proposal-json-modules, or add a new position entry for JSON modules?

In case it's relevant, JSON modules is now a Stage 3 proposal (as is import assertions).

Thanks!

**Comment by codehag** [comment]
Hi @dandclark, In this case, I don't think we need to explicitly update it, since I gave support for it at TC39 for stage 3 and that was mozilla's position both then and now. We have started implementing both. 

cc @tantek

**Comment by dandclark** [comment]
Ok, thanks @codehag ! It's great to hear that the implementation is underway.

**Comment by danielweil** [comment]
Any updates on this issue implementation?

**Comment by ekr** [comment]
This is not the place to look updates on implementation in Gecko. Perhaps @codehag can add the bugzilla link?

**Comment by codehag** [comment]
Import assertions is finished https://bugzilla.mozilla.org/show_bug.cgi?id=1736059

As it is still at stage 3, we haven't shipped it unflagged yet, but the flag for enabling it is `experimental.import_assertions`. We have an external contributor working on JSON modules, which is the first "assertion" that will be exposed. You can ask further questions on https://bugzilla.mozilla.org/show_bug.cgi?id=1670176.

**Comment by 321paranoiawhy** [comment]
Hi, I can not use import assertions in `FireFox` version `103.0.2 (64 位)`.

**Comment by myakura** [comment]
@321paranoiawhy it's not shipped yet https://bugzilla.mozilla.org/show_bug.cgi?id=1777526

**Comment by juner** [comment]
@321paranoiawhy Try using the following code.
```js
async function enableWithTypeSyntax() {
  const option = (() => {
    const option = {};
    const w = {type:"json"};
    Reflect.defineProperty(option, "with", {
      get: () => (enableWithTypeSyntax+=1,w),
    });
    return option;
  })();
  let enableWithTypeSyntax = 0;
  let result;
  try {
    result = await import("data:application/json;base64,e30=", option);
  }catch(e){}
  return enableWithTypeSyntax>0;
}
```

[playground](https://v26.livecodes.io/?x=code/N4IgLglmA2CmIC4QFUB2kawCYAIAKATgPYBWsAxmCADQhawDO5BEADpEaoiDSABawAhlm4AeALawwgnOT6CCDKQF4AOigAqAMQC0ADnU4A9AD5VqCVJmpBktSABuEWAHdWRAmEPlOYWOnsXCCwwPmV6J3JYHSCQvmocCFQoCEFoHSY02GUARgA6AAZDU14+MHFoAEEwMEVuaEFUAHN7f28GhgZ7dV5pJoZEAG0AXVpBSggHWABRLCgPbnEFAGsAV1ZeJYI1jYRQBubVwSb4JDKK3h90fyokUTmHEwhxd08cQRqWACNVvxwGdavMAIHCiBisRqJLD2AiMVbQLwgEyoIhgf6AjxgURGcGNEzYh5mAjmMHMNhosAAT1Y2XU4iIWHhsHURPMOHeDEpqHIOAAZqtuRxUDh-IIvnAAOpQPgaamwADKXOkAA8ABQAShwwFUxOF7NknAYaKI7AgnBwyhwqo1FpMWp1bP17KuRpwJqFFq1AF8ANwOvVOl1olye4BUmkIdQkBicdS+-1O9kAJVgvLglDy9F5SVghBNsE8lNV7rNqAS6iCoXUCW1usT+pOwKtmuUdtVovFsCloVlNMV6EEyoA1LlqC51dQE4mveq-XXE7CwKsCMKS5w5472TONRuAzg4GiO5Lpb2FUrB56CrunQecLCAQjr-rapT7fOnff4WjLYIXIIoIkLyYqq6hYB8ggIIIrCsNAEDkB8pZGNG65fIISgAGwACzULAADMBT2Aka6oLOU5bvBYByO26rAF6ZF3lIy7CkeXYnnK-YqiYV5TnR863gC0GYp6v7-oeNidt2MrseeaqkfOLpEHAeTQEQTSqgJQJyY6n4InkSSoAWGiwMq344AABgAJMAGlCQA-Dg6g2Z4hgguoKJok5iJemZu7YkwLDsCYIBerQRqUnAiD7I0TRHCc3DkJ0ly+Dc3DBaFZLsJFIAHDFxynCAJCCA4aEZVQtBXH46CpSFIBhXADACFIAwICM6UBWAzWtSACUMIQjBSKltCsMQUSdB4nWjN1qxGkQ4jylIkDNM1tG0M8QLLTV4aMJFm2MB1WU5bF+Vbf55JJdcVVIGljgFgwpbcAATOhvCwsInDQJSiC8mkShjNAKkuAAMtFADC8jNKctSrLAtD0vQ3D8v9vREIpy0gCxIhIGk0C8OMkBTPFhqKfAoXSEuzXdSpSgiDVABeKPiIgORekAA)
